### PR TITLE
Add support for Commandline Mode  to the CmdPal

### LINKED ISSUE #6677: Command Palette: Add support for commandline mode
Follow up to #6635.

Users should be able to type `> new-tab` to run a `new-tab` command. This is a simple example - more complex commandlines are also supported. 

This is currently blocked on:
* [x] #6635 
* [x] #6537 

There are more details in the spec.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Adds support for "commandline mode" to the command palette. 
cmdpal-commandline-mode

This allows the user to start typing a `wt.exe` commandline directly in the command palette, to run that commandline directly in the current window. This allows the user input something like `> nt -p Ubuntu ; sp -p ssh` and open up a new tab and split it _in the current window_. 

## References

* cmdpal megathread: #5400
* Kinda related to #4472
* built with the `wt` action from #6537

## PR Checklist
* [x] Closes #6677
* [x] I work here
* [ ] Tests added/passed
* [ ] Requires documentation to be updated - sure does, when the cmdpal docs are written in the first place :P

## Validation Steps Performed

Tested manually

### CODE REVIEW

**Comment by WSLUser** [comment]
Can we get auto-hinting/Intellisense as you type? (This could be a follow-on PR).

**Comment by zadjii-msft** [comment]
So "intellisense" might be impossible with our current arg parser. We'd probably have to roll our own parser, and I'm not a compiler expert, and wouldn't have any idea how to start that. 

What I am planning on is showing any errors from the parsing in the palette, in #7284

**Comment by WSLUser** [comment]
Maybe the Visual Studio/VS Code C++ extension folks have some ideas either way.

**Comment by DHowett** [comment]
The.. people who write the typescript package that integrates C++ with VS Code might .. be able to help us with .. commandline parsing? 🤷

**Comment by WSLUser** [comment]
They provide Intellisense to the editor. Maybe a similar strategy can be adopted?

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 1]
```diff
@@ -181,23 +181,39 @@ namespace winrt::TerminalApp::implementation
...
+                {
+                    _searchBox().Text(L"");
+                }
             }
-            else
+            else if (_currentMode == CommandPaletteMode::CommandlineMode)
             {
-                _searchBox().Text(L"");
+                // CommandlineMode Mode: always exit the palette immediately.
+                _dismissPalette();
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.cpp:216]: Shouldn't this be consistent with "ESC once to clear, twice to exit"? Long-term, switching modes is supposed to be which prefix we have, so they should line up right?

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 2]
```diff
@@ -408,12 +466,36 @@ namespace winrt::TerminalApp::implementation
...
+            _checkActionVsCommandlineMode();
+        }
+
         _updateFilteredActions();
         _filteredActionsView().SelectedIndex(0);
 
         _noMatchesText().Visibility(_filteredActions.Size() > 0 ? Visibility::Collapsed : Visibility::Visible);
     }
 
+    void CommandPalette::_checkActionVsCommandlineMode()
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.cpp:480]: nit: could we rename this to "_evaluatePrefix" or something more extensible? As we add more modes, I wouldn't want to rename this "_checkActionVsCommandlineVsProfileSelectorVsTabSwitcherMode" haha

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/CommandPalette.cpp:480]: Sure why not, I hated that method name anyways

File: src/cascadia/TerminalApp/CommandPalette.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/CommandPalette.cpp:216]: Yea okay. I can work with that.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 3]
```diff
@@ -422,6 +458,72 @@ namespace winrt::TerminalApp::implementation
...
+    void CommandPalette::_dispatchCommandline()
+    {
+        const auto input = _getPostPrefixInput();
+        if (input.empty())
+        {
+            return;
+        }
+        winrt::hstring cmdline{ input };
+
+        // Build the NewTab action from the values we've parsed on the commandline.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/CommandPalette.cpp:507]: Whoops yea, good catch. It's definitely _not_ haha

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 4]
```diff
@@ -174,23 +174,57 @@ namespace winrt::TerminalApp::implementation
...
+                _dispatchCommandline();
             }
 
             e.Handled(true);
         }
         else if (key == VirtualKey::Escape)
         {
-            // Action Mode: Dismiss the palette if the text is empty, otherwise clear the search string.
-            if (_searchBox().Text().empty())
+            if (_currentMode == CommandPaletteMode::ActionMode)
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/CommandPalette.cpp:193]: Thanks, good catches on code that predated the tab* modes 😆

**Comment by zadjii-msft** [comment]
This PR is currently blocked on the CI issue fixed over in #7383 

@msftbot merge this in 5 minutes

**Pull request merged**